controller/src/main/java/io/pravega/controller/server/ControllerServiceMain.java

```diff
@@ -146,6 +144,10 @@ protected void run() throws Exception {
 
                 log.info("Awaiting termination of ControllerServiceStarter");
                 starter.awaitTerminated();
+
+                if (hasZkConnection) {
```

Comment:
Do you need to do this? You already have this in the `finally` block

Reply:
This is inside the `while(isRunning)` loop where we close and then create all modules again including a new curator client.

the try-catch-finally is overall wrapper over the while loop in case an unexpected exception is thrown.

Reply:
If the session has expired, then aren't we supposed to be leaving the loop?